Add CMerkleTx::IsImmatureCoinBase method

Empact opened the pull request:
All but one call to `GetBlocksToMaturity` is testing it relative to 0
for the purposes of determining whether the coinbase tx is immature.
In such case, the value greater than 0 implies that the tx is coinbase,
so there is no need to separately test that status.

This names the concept for easy singular use.

Empact committed:
Add CMerkleTx::IsImmatureCoinBase method

All but one call to GetBlocksToMaturity is testing it relative to 0
for the purposes of determining whether the coinbase tx is immature.
In such case, the value greater than 0 implies that the tx is coinbase,
so there is no need to separately test that status.

This names the concept for easy singular use.

251Labs commented on src/wallet/wallet.cpp:
```diff
@@ -1924,7 +1924,7 @@ CAmount CWalletTx::GetDebit(const isminefilter& filter) const
 {
     // Must wait until coinbase is safely deep enough in the chain before valuing it
-    if (IsCoinBase() && GetBlocksToMaturity() > 0)
+    if (IsImmatureCoinBase())
```
It seems safe to optimize this boolean expression and remove the `IsCoinBase()` condition in `IsImmatureCoinBase`, because `GetBlocksToMaturity` already calls `IsCoinBase`, which effectively makes this condition redundant. 

I do however appreciate the defensiveness of having the redundant`IsCoinBase()` condition; it is not specified that `GetBlocksToMaturity` must return `0` if it is not a coinbase transaction.

Therefore, if we optimize the boolean expression in `IsImmatureCoinBase` and fully rely on the result of `GetBlocksToMaturity`, I would like to suggest to add a comment to the interface documentation of the `CMerkleTx::GetBlocksToMaturity` method that explicitly states the postcondition that `GetBlocksToMaturity` must return `0` if it is not a coinbase transaction.

251Labs commented on src/wallet/wallet.h:
```diff
@@ -277,6 +277,8 @@ class CMerkleTx
     const uint256& GetHash() const { return tx->GetHash(); }
     bool IsCoinBase() const { return tx->IsCoinBase(); }
+    // note GetBlocksToMaturity is 0 for non-coinbase tx
+    bool IsImmatureCoinBase() const { return GetBlocksToMaturity() > 0; }
```
See my first comment about relying on the result of `GetBlocksToMaturity`.

l2a5b1 reviewed (commented):
Nice refactor, utACK.

promag commented:
Concept ACK. Makes the intention clear. Agree with @251Labs points.

Empact commented:
Added docs, moved the implementations together.

I also noticed that the result of `GetBlocksToMaturity` would be inaccurate if the TX was marked as conflicted - my impression is that coinbase transactions can't be conflicting, so I added an assert to make that expectation explicit:
https://github.com/bitcoin/bitcoin/pull/13631/files#diff-b2bb174788c7409b671c46ccc86034bdR4440

Empact commented:
Moved the GetBlocksToMaturity assert out into #13657

Empact commented:
Rebased for #13630, #13072

251Labs commented:
Nice, utACK 860e214

promag commented:
utACK 860e214.

MarcoFalke commented:
utACK 860e214f7ba899efae397205891181056adf3fc2

Empact commented:
Rebased for #12257

MarcoFalke commented:
re-utACK bb653872be8d251c21ee32c2948100b7febbd477

Empact commented:
Noticed the whitespace was off. Diff:

```diff
diff --git a/src/wallet/wallet.cpp b/src/wallet/wallet.cpp
index fc6f03a16..540a7b0fc 100644
--- a/src/wallet/wallet.cpp
+++ b/src/wallet/wallet.cpp
@@ -1926,8 +1926,7 @@ CAmount CWalletTx::GetCredit(const isminefilter& filter) const
 
 CAmount CWalletTx::GetImmatureCredit(bool fUseCache) const
 {
-    if (IsImmatureCoinBase() && IsInMainChain())
-    {
+    if (IsImmatureCoinBase() && IsInMainChain()) {
         if (fUseCache && fImmatureCreditCached)
             return nImmatureCreditCached;
         nImmatureCreditCached = pwallet->GetCredit(*tx, ISMINE_SPENDABLE);
@@ -1985,8 +1984,7 @@ CAmount CWalletTx::GetAvailableCredit(bool fUseCache, const isminefilter& filter
 
 CAmount CWalletTx::GetImmatureWatchOnlyCredit(const bool fUseCache) const
 {
-    if (IsImmatureCoinBase() && IsInMainChain())
-    {
+    if (IsImmatureCoinBase() && IsInMainChain()) {
         if (fUseCache && fImmatureWatchCreditCached)
             return nImmatureWatchCreditCached;
         nImmatureWatchCreditCached = pwallet->GetCredit(*tx, ISMINE_WATCH_ONLY);
@@ -4399,8 +4397,8 @@ int CMerkleTx::GetBlocksToMaturity() const
 
 bool CMerkleTx::IsImmatureCoinBase() const
 {
-   // note GetBlocksToMaturity is 0 for non-coinbase tx
-   return GetBlocksToMaturity() > 0;
+    // note GetBlocksToMaturity is 0 for non-coinbase tx
+    return GetBlocksToMaturity() > 0;
 }
 
 bool CWalletTx::AcceptToMemoryPool(const CAmount& nAbsurdFee, CValidationState& state)
```

laanwj commented:
utACK 23f434378153cf764230066662f3ec3ad614ff30
